[docs] Add transitions customization page (#21456)
mui-org/material-ui: docs/src/pages/customization/transitions/transitions.md

# Transitions

> <p class="description">Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>

Reviewer: the generated HTML is:

```html
<meta name="description" content="Use the &lt;code&gt;theme.transitions.create()&lt;/code&gt; helper to create consistent transitions on the elements of your UI."/>
```

I'm curious to see what it looks like once indexed. We will see :)

> {{"demo": "pages/customization/transitions/TransitionHover.js"}}

Reviewer: I don't think that the inline preview is of any value. We can disable it to reduce noise.

```suggestion
{{"demo": "pages/customization/transitions/TransitionHover.js", "defaultCodeOpen": false}}
```

Reviewer: Or we could do the opposite, open it completely
```suggestion
{{"demo": "pages/customization/transitions/TransitionHover.js", "defaultCodeOpen": true}}
```

Author: Let me change the example to see how long it would be and I will check what will look better

# Transitions

<p class="description">Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>

> 
> {{"demo": "pages/customization/transitions/TransitionHover.js", "defaultCodeOpen": false}}

Reviewer: I think that it could be interesting to have a visual example people can directly copy & paste.

```suggestion
```jsx
theme.transitions.create(['background-color', 'transform']);
```

{{"demo": "pages/customization/transitions/TransitionHover.js", "defaultCodeOpen": false}}
```

Author: Sry I don’t understand what is the suggested change here. Should I add code snippet after the example?

Author: Ah I see the code snippet is added before the example :)

## Custom duration

You can change some of the durations by providing a number.

```js
const theme = createMuiTheme({
  transitions: {
    duration: {
      standard: 500,
    },    
  },
});
```

## Custom easing

You can change some of the easing values by providing a custom <code>transition-timing-function</code> value.

```js
const theme = createMuiTheme({
  transitions: {
    easing: {
      sharp: 'cubic-bezier(0.4, 0, 0.8, 1)',
    },    
  },
});
```

## API

### `theme.transitions.create(props, options) => transition`

#### Arguments

> 1. `props` (*String* | *String[]*): Defaults to `[all]`. Provides a css property, or a list of css properties that should be transitioned.

Reviewer: ```suggestion
1. `props` (*String* | *String[]*): Defaults to `['all']`. Provides a css property, or a list of css properties that should be transitioned.
```

# Transitions

> <p class="description">Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>

Author: Ok, let me restructure it a bit, by starting with the customization options for the transitions (and mentioning that they will affect the components that use them). Then I will move to the `duration`, `easing`, `getAutoHeightDuration` and in the end the `create` API.

For the `getAutoHeightDuration` is this description good:
- "Returns the automatic duration based on the provided height"

> <p class="description">Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>

Author: Address all comments, please take another look

> ```jsx

Reviewer: ```suggestion
```js
```

theme.transitions.create(['background-color', 'transform']);
```

{{"demo": "pages/customization/transitions/TransitionHover.js", "defaultCodeOpen": false}}

## Custom duration

You can change some of the durations by providing a number.

```js
const theme = createMuiTheme({
  transitions: {

>     duration: {
>       standard: 500,
>     },    

Reviewer: I think that we can mention all the possible values

```suggestion
    duration: {
      shortest: 150,
      shorter: 200,
      short: 250,
      // most basic recommended timing
      standard: 300,
      // this is to be used in complex animations
      complex: 375,
      // recommended when something is entering screen
      enteringScreen: 225,
      // recommended when something is leaving screen
      leavingScreen: 195,
    },
```

  },
});
```

## Custom easing

You can change some of the easing values by providing a custom <code>transition-timing-function</code> value.

```js
const theme = createMuiTheme({
  transitions: {
    easing: {

>       sharp: 'cubic-bezier(0.4, 0, 0.8, 1)',

Reviewer: I think that we can mention all the easing curves:

```suggestion
      // This is the most common easing curve.
      easeInOut: 'cubic-bezier(0.4, 0, 0.2, 1)',
      // Objects enter the screen at full velocity from off-screen and
      // slowly decelerate to a resting point.
      easeOut: 'cubic-bezier(0.0, 0, 0.2, 1)',
      // Objects leave the screen at full velocity. They do not decelerate when off-screen.
      easeIn: 'cubic-bezier(0.4, 0, 1, 1)',
      // The sharp curve is used by objects that may return to the screen at any time.
      sharp: 'cubic-bezier(0.4, 0, 0.6, 1)',
```

# Transitions

<p class="description">Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>

```js
theme.transitions.create(['background-color', 'transform']);
```

{{"demo": "pages/customization/transitions/TransitionHover.js", "defaultCodeOpen": false}}

## Custom duration

> You can change some of the durations by providing a number.

Reviewer 2: ```suggestion
You can change some or all of the duration values, or provide your own (for use in the `create()` helper). This example uses the default values which are in milliseconds. You only need to provide the keys you wish to change or add a value for. 
```

(The last sentence might be stating the obvious, but you'd be surprised!)

```js
const theme = createMuiTheme({
  transitions: {
    duration: {
      shortest: 150,
      shorter: 200,
      short: 250,
      // most basic recommended timing
      standard: 300,
      // this is to be used in complex animations
      complex: 375,
      // recommended when something is entering screen
      enteringScreen: 225,
      // recommended when something is leaving screen
      leavingScreen: 195,
    },
  },
});
```

## Custom easing

> You can change some of the easing values by providing a custom <code>transition-timing-function</code> value.

Reviewer 2: ```suggestion
You can change some or all of the easing values, or provide your own, by providing a custom CSS <code>transition-timing-function</code> value.
```

## API

### `theme.transitions.create(props, options) => transition`

#### Arguments

1. `props` (*String* | *String[]*): Defaults to `['all']`. Provides a CSS property, or a list of CSS properties that should be transitioned.
2.  `options` (*Object* [optional]):
  - `options.duration` (*String* | *Number* [optional]): Defaults to `theme.transitions.duration.standard`. Provides the duration of the transition.
  - `options.easing` (*String* [optional]): Defaults to `theme.transitions.easing.easeInOut`. Provides the easing for the transition.
  - `options.delay` (*String* | *Number* [optional]): Defaults to `0`. Provides the delay for the transition.

#### Returns

`transition`: A transition CSS value, which composes all properties which should be transitioned, together with the defined duration, easing and duration.

Use the <code>theme.transitions.create()</code> helper to create consistent transitions for the elements of your UI.</p>

```js
theme.transitions.create(['background-color', 'transform']);
```

> 

Reviewer 2: ```suggestion

#### Example

```

2.  `options` (*Object* [optional]):
  - `options.duration` (*String* | *Number* [optional]): Defaults to `theme.transitions.duration.standard`. Provides the duration of the transition.
  - `options.easing` (*String* [optional]): Defaults to `theme.transitions.easing.easeInOut`. Provides the easing for the transition.
  - `options.delay` (*String* | *Number* [optional]): Defaults to `0`. Provides the delay for the transition.

#### Returns

`transition`: A transition CSS value, which composes all properties which should be transitioned, together with the defined duration, easing and duration.

Use the <code>theme.transitions.create()</code> helper to create consistent transitions for the elements of your UI.</p>

```js
theme.transitions.create(['background-color', 'transform']);
```

{{"demo": "pages/customization/transitions/TransitionHover.js", "defaultCodeOpen": false}}

### `theme.transitions.getAutoHeightDuration(height) => duration`

#### Arguments

1. `height` (*Number*): The height of the component.

#### Returns

`duration`: The calculated duration based on the height.

## Durations

> You can change some or all of the duration values, or provide your own (for use in the `create()` helper). This example uses the default values which are in milliseconds. You only need to provide the keys you wish to change or add a value for.

Reviewer 2: ```suggestion
You can change some or all of the duration values, or provide your own for use in the `create()` helper. This example uses the default values, which are in milliseconds. You only need to provide the keys you wish to change or add.
```

Reviewer 2: Still not totally happy with that, but I think it's a little better than the previous rewording.

## Custom duration

> You can change some or all of the duration values, or provide your own (for use in the `create()` helper). This example uses the default values which are in milliseconds. You only need to provide the keys you wish to change or add a value for. 

Reviewer 2: ```suggestion
You can change some or all of the duration values, or provide your own (for use in the `create()` helper). This example uses the default values which are in milliseconds. You only need to provide the keys you wish to change or add a value for.
```
Oops, sorry, trailing space.

});
```

## `getAutoHeightDuration`

> You can customize the default function for calculating the duration based on the height of the elments (useful for collapsing/expanding) components.

Reviewer 2: ```suggestion
You can customize the default function for calculating the duration based on the height of the element (useful for collapsing/expanding transitions).
```

```js
const theme = createMuiTheme({
  transitions: {
    getAutoHeightDuration: (height) => {
      if (!height) {
        return 0;
      }

      const constant = height / 36;

      return Math.round((4 + 15 * constant ** 0.25 + constant / 5) * 10);
    },
  },
})
```

## `create`

> Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>

Reviewer 2: ```suggestion
Use the <code>theme.transitions.create()</code> helper to create consistent transitions for the elements of your UI.</p>
```
